File: verdeciel/Icon.swift
//  Created by Devine Lu Linvega on 2015-09-21.
//  Copyright (c) 2016 XXIIVV. All rights reserved.

import UIKit
import QuartzCore
import SceneKit
import Foundation

class Icon : Empty
{
	var host:Location!
	
	var color:UIColor = UIColor.purpleColor()
	var size:Float = 0.1
	
	var label = SCNLabel()
	var trigger = Empty()
	var mesh = Empty()
	var wire:SCNLine!
	
	override init()
	{
		super.init()
		
		label = SCNLabel(text: "", scale: 0.06, align: alignment.center, color: grey)
		label.position = SCNVector3(0,-0.3,-0.35)
		self.addChildNode(label)
		
		wire = SCNLine(vertices: [],color:white)
		wire.position = SCNVector3(0,0,-0.01)
		wire.hide()
		self.addChildNode(wire)
		
		addChildNode(mesh)
		addChildNode(label)
		addChildNode(trigger)
		addChildNode(wire)
	}
	
	func addHost(host:Location)
	{
		self.host = host
		
		label.update(host.name!)
	}
	
	override func whenStart()
	{
		super.whenStart()
		if host.mapRequirement != nil { label.update(cyan) }
	}
	
	func onUpdate()
	{
		if host.isComplete == nil { color = white }
		else if host.isComplete == false { color = red }
		else if host.isComplete == true { color = cyan }
		
		mesh.updateChildrenColors(color)
	}
	
	func close()
	{
	}
	
	required init?(coder aDecoder: NSCoder)
	{
		fatalError("init(coder:) has not been implemented")
	}
}

File: verreciel_js/scripts/objects/icon.js
class Icon extends Empty
{
  constructor()
  {
    super();

    this.color = new THREE.Vector4(0.5, 0, 0.5, 1); // Haha, weird!
    this.size = 0.1;
    
    this.trigger = new Empty();
    this.mesh = new Empty();
    this.label = new SceneLabel("", 0.06, Alignment.center, verreciel.grey);
    this.label.position.set(0,-0.3,-0.35);
    this.add(this.label);
    
    this.wire = new SceneLine([], verreciel.white);
    this.wire.position.set(0,0,-0.01);
    this.wire.hide();
    this.add(this.wire);
    
    this.add(this.mesh);
    this.add(this.label);
    this.add(this.trigger);
    this.add(this.wire);
  }
  
  addHost(host)
  {
    this.host = host;
    this.label.update(this.host.name);
  }
  
  whenStart()
  {
    super.whenStart();
    if (this.host.mapRequirement != null)
    {
      this.label.update(verreciel.cyan);
    }
  }
  
  onUpdate()
  {
    if (this.host.isComplete == null)
    {
      this.color = verreciel.white;
    }
    else if (this.host.isComplete == false)
    {
      this.color = verreciel.red;
    }
    else if (this.host.isComplete == true)
    {
      this.color = verreciel.cyan;
    }
    
    this.mesh.updateChildrenColors(this.color);
  }
  
  close()
  {
  }
}
